test-data/unit/check-narrowing.test

```diff
@@ -445,3 +528,225 @@ if y["model"]["key"] is Key.C:
+x3: Foo
+if x3 == A_literal:
+    reveal_type(x3)  # N: Revealed type is 'Literal[__main__.Foo.A]'
+else:
+    reveal_type(x3)  # N: Revealed type is 'Literal[__main__.Foo.B]'
+[builtins fixtures/primitives.pyi]
+
+[case testNarrowingEqualityDisabledForCustomEquality]
+from typing import Union
+from typing_extensions import Literal
+from enum import Enum
+
+class Custom:
+    def __eq__(self, other: object) -> bool: return True
+
+class Default: pass
+
+x1: Union[Custom, Literal[1], Literal[2]]
+if x1 == 1:
+    reveal_type(x1)  # N: Revealed type is 'Union[__main__.Custom, Literal[1], Literal[2]]'
+else:
+    reveal_type(x1)  # N: Revealed type is 'Union[__main__.Custom, Literal[1], Literal[2]]'
+
+x2: Union[Default, Literal[1], Literal[2]]
+if x2 == 1:
+    reveal_type(x2)  # N: Revealed type is 'Literal[1]'
+else:
+    reveal_type(x2)  # N: Revealed type is 'Union[__main__.Default, Literal[2]]'
+
+class CustomEnum(Enum):
+    A = 1
+    B = 2
+
+    def __eq__(self, other: object) -> bool: return True
+
+x3: CustomEnum
+key: Literal[CustomEnum.A]
+if x3 == key:
+    reveal_type(x3)  # N: Revealed type is '__main__.CustomEnum'
+else:
+    reveal_type(x3)  # N: Revealed type is '__main__.CustomEnum'
+
+# For comparison, this narrows since we bypass __eq__
+if x3 is key:
+    reveal_type(x3)  # N: Revealed type is 'Literal[__main__.CustomEnum.A]'
+else:
+    reveal_type(x3)  # N: Revealed type is 'Literal[__main__.CustomEnum.B]'
+[builtins fixtures/primitives.pyi]
+
+[case testNarrowingEqualityDisabledForCustomEqualityChain]
+# flags: --strict-optional --strict-equality --warn-unreachable
+from typing import Union
+from typing_extensions import Literal
+
+class Custom:
+    def __eq__(self, other: object) -> bool: return True
+
+class Default: pass
+
+x: Literal[1, 2, None]
+y: Custom
+z: Default
+
+# We could maybe try doing something clever, but for simplicity we
+# treat the whole chain as contaminated and mostly disable narrowing.
+#
+# The only exception is that we do at least strip away the 'None'. We
+# (perhaps optimistically) assume no custom class would be pathological
+# enough to declare itself to be equal to None and so permit this narrowing,
+# since it's often convenient in practice.
+if 1 == x == y:
+    reveal_type(x)   # N: Revealed type is 'Union[Literal[1], Literal[2]]'
+    reveal_type(y)   # N: Revealed type is '__main__.Custom'
+else:
+    reveal_type(x)   # N: Revealed type is 'Union[Literal[1], Literal[2], None]'
+    reveal_type(y)   # N: Revealed type is '__main__.Custom'
+
+# No contamination here
+if 1 == x == z:      # E: Non-overlapping equality check (left operand type: "Union[Literal[1], Literal[2], None]", right operand type: "Default")
+    reveal_type(x)   # E: Statement is unreachable
+    reveal_type(z)
+else:
+    # TODO: Why do we narrow away 'Literal[1]' here?
+    # Even if the equality comparison is bogus, we should try and do better here.
```

Comment:
This is a bit worrying, so I think this deserves a follow-up issue.

Reply:
Well, this is a bit embarrassing. I did some more digging to try and it turned out that the whole concept of "partial TypeMaps" I introduced in my earlier diff was incorrect and that several of the narrowing test cases I added were too optimistic.

Basically, I incorrectly made mypy narrow the else case in programs like:

```
x: Literal[1, 2]
y: Literal[1, 2]
if x == 2 == y:
    reveal_type(x)  # Revealed type is Literal[2]
    reveal_type(y)  # Revealed type is Literal[2]
else:
    reveal_type(x)  # Revealed type is incorrectly Literal[1]!
    reveal_type(y)  # Revealed type is incorrectly Literal[1]!
```

But actually, we could enter the else case if x == 2 and y == 1 or vice versa, so narrowing was incorrect.

Going back to using just `or_conditional_maps` instead of `or_partial_conditional_maps` ended up fixing these bogus narrowings. I also added a few more test cases + fixed some earlier enum ones.